homework/philosopher.cpp

```diff
@@ -0,0 +1,144 @@
+#include <algorithm>    // std::for_each()
+#include <atomic>
+#include <chrono>
+#include <cstdlib>
+#include <ctime>
+#include <functional>   // std::mem_fn() 
+#include <iostream>
+#include <mutex>
+#include <string>
+#include <sstream>
+#include <thread>
+
+constexpr size_t itemsCnt = 5;       // number of items (philosophers, forks)
+constexpr size_t actionCnt = 10;     // number of action to take by philosopher
+constexpr bool action_debug = false; // additional debugs
+
+std::mutex forks[ itemsCnt ];
+std::thread philosopher[itemsCnt];
+
+std::mutex displayMutex;
+std::mutex updateMutex;
+std::atomic<bool> canEat[itemsCnt];
+std::atomic<bool> eat[itemsCnt];
+
+using namespace std::chrono_literals;
+
+void displayState()
+{
+    std::stringstream ss;
+
+    std::lock_guard<std::mutex> lock( displayMutex );
```

Comment:
Przy użyciu string streama w ogóle nie trzeba blokować wypisywania na ekran. Na pewno będzie ono niepodzielne, bo do cout robisz tylko jedną operację zapisu.

Reply:
Racja co do string streama. Natomiast zadaniem updateMutex jest to aby blokowac zmiane tablicy eat[] na czas wyswietlania.
W innym przypadku od czasu do czasu powstaja kwiatki w rodzaju: -F0- -F1- F2 F3 F4.